src/language_features/hover.rs

```diff
@@ -109,17 +111,36 @@ pub fn editor_hover(
         },
     };
 
-    if contents.is_empty() && diagnostics.is_empty() {
+    if contents.is_empty() && diagnostics.is_empty() && maybe_hover_modal.is_none() {
         return;
     }
 
-    let command = format!(
+    let anchor_or_style = if let Some(hover_modal) = maybe_hover_modal.as_ref() {
+        contents = format!("{}\n---\n{}", hover_modal.context, contents);
+        "modal".to_string()
+    } else {
+        format!("{}", params.position)
+    };
+
+    let mut command = format!(
         "lsp-show-hover {} %§{}§ %§{}§",
-        params.position,
+        anchor_or_style,
         contents.replace("§", "§§"),
         diagnostics.replace("§", "§§")
     );
 
+    // Wrap if we're using a HoverModal
+    if let Some(hover_modal) = maybe_hover_modal {
+        command = format!(
+            "eval %§
+                 {}
```

Comment:
The extra indentation in continuation lines will look weird in the log, so I think it's better to use

```rust
            "eval %§\n    {}\n    {}\n§",
```

not sure if we use this style of indentation elsewhere but why not I guess.

Same thing happens a few more times in this patch. I wonder if it's worth looking into a general solution. Sadly I haven't found a conventional way to do this in Rust.

Java, out of all languages gets multiline string literals right - unlike most other languages they automatically remove common indentation in strings like

```java
	String x = """
		123
			456
		789
	""";
```

This will evaluate to the same as `"123\n\t456\789\n"` (not sure about leading/trailing whitespace, but that's not the point).
Maybe we should add a function that does that, should be very simple.

Reply:
Looks like the [indoc!](https://crates.io/crates/indoc) macro would solve this. Any concerns with adding that dependency?

Reply:
Done with 0cc20e9 (Use indoc for multiline string literals) - (sorry for the force-push), so this can be

```diff
diff --git a/src/language_features/hover.rs b/src/language_features/hover.rs
index 149f9b6..3ccf100 100644
--- a/src/language_features/hover.rs
+++ b/src/language_features/hover.rs
@@ -2,6 +2,7 @@ use crate::context::*;
 use crate::markup::*;
 use crate::types::*;
 use crate::util::*;
+use indoc::formatdoc;
 use itertools::Itertools;
 use lsp_types::request::*;
 use lsp_types::*;
@@ -131,8 +132,8 @@ pub fn editor_hover(
 
     // Wrap if we're using a HoverModal
     if let Some(hover_modal) = maybe_hover_modal {
-        command = format!(
-            "eval %§
+        command = formatdoc!(
+            "evaluate-commands %§
                  {}
                  {}
              §",
diff --git a/src/language_features/kak_lsp_custom.rs b/src/language_features/kak_lsp_custom.rs
index dc27f59..df77ba8 100644
--- a/src/language_features/kak_lsp_custom.rs
+++ b/src/language_features/kak_lsp_custom.rs
@@ -1,6 +1,7 @@
 use crate::{
     context::*, language_features::hover::editor_hover, markup::escape_brace, types::*, util::*,
 };
+use indoc::{formatdoc, indoc};
 use lsp_types::{request::*, *};
 use serde::Deserialize;
 use url::Url;
@@ -113,16 +114,17 @@ fn next_or_prev_response(
                             );
 
             // This kak script is appended after the kakoune `lsp-show-hover`
-            let do_after = format!(
-                "on-key %[eval %sh[
-                     if [ \"$kak_key\" = \"g\" ];
-                     then echo 'info -style modal
-                                exec {}g{}lh
-                                ';
-                     else echo 'info -style modal';
-                     fi
-                 ]]",
-                symbol_position.line, symbol_position.column
+            let do_after = formatdoc!(
+                "on-key %[
+                     info -style modal
+                     evaluate-commands %sh[
+                         if [ \"$kak_key\" = \"g\" ]; then
+                             echo 'exec {}g{}lh'
+                         fi
+                     ]
+                 ]",
+                symbol_position.line,
+                symbol_position.column
             );
 
             // Make a HoverRequest to the backend language server and then call `hover::editor_hover()`
@@ -154,10 +156,14 @@ fn next_or_prev_response(
     } else {
         ctx.exec(
             meta,
-            "eval %[
-                info -style modal 'Not found!\n\nPress any key to continue'
-                on-key %[info -style modal]
-            ]",
+            indoc!(
+                "evaluate-commands %[
+                     info -style modal 'Not found!
+
+                     Press any key to continue'
+                     on-key %[info -style modal]
+                 ]"
+            ),
         );
     }
 }
```